File: .claude/skills/crucible-spec-workspace/grade_spec.py
#!/usr/bin/env python3
"""Grade crucible-spec outputs against assertions."""
import json
import re
import sys
from pathlib import Path


def grade_spec(spec_path: str, metadata_path: str) -> dict:
    spec = Path(spec_path).read_text(encoding="utf-8") if Path(spec_path).exists() else ""
    metadata = json.loads(Path(metadata_path).read_text(encoding="utf-8"))

    results = []
    for assertion in metadata["assertions"]:
        name = assertion["name"]
        text = assertion["text"]
        passed = False
        evidence = ""

        if assertion.get("type") == "regex":
            pattern = assertion["pattern"]
            matches = re.findall(pattern, spec)
            passed = len(matches) > 0
            evidence = f"Found {len(matches)} matches: {matches[:3]}" if passed else f"No match for pattern: {pattern}"

        elif assertion.get("type") == "heading_exists":
            heading = assertion["heading"]
            # Check for markdown headings containing the keyword (case-insensitive)
            pattern = rf"^#{{1,4}}\s+.*{re.escape(heading)}.*$"
            matches = re.findall(pattern, spec, re.MULTILINE | re.IGNORECASE)
            passed = len(matches) > 0
            evidence = f"Found heading: {matches[0]}" if passed else f"No heading containing '{heading}'"

        elif assertion.get("type") == "content_check":
            # Special case for notification types
            if "알림 타입" in text or "notification" in text.lower():
                keywords = ["주문", "메시지", "공지"]
                found = [k for k in keywords if k in spec]
                passed = len(found) >= 3
                evidence = f"Found {len(found)}/3 notification types: {found}"

        elif assertion.get("type") == "composite":
            # gate-spec passable: check all required sections
            required = ["Problem", "Journey", "Feature", "Constraint", "Acceptance", "Data Model"]
            found = []
            missing = []
            for req in required:
                if re.search(rf"^#{{1,4}}\s+.*{req}.*$", spec, re.MULTILINE | re.IGNORECASE):
                    found.append(req)
                else:
                    missing.append(req)
            passed = len(missing) == 0
            evidence = f"Found {len(found)}/{len(required)} sections. Missing: {missing}" if missing else f"All {len(required)} required sections present"

        results.append({"text": text, "passed": passed, "evidence": evidence})

    return {
        "eval_id": metadata["eval_id"],
        "eval_name": metadata["eval_name"],
        "expectations": results,
        "pass_rate": sum(1 for r in results if r["passed"]) / len(results) if results else 0,
    }


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: grade_spec.py <spec_file> <metadata_file>")
        sys.exit(1)
    result = grade_spec(sys.argv[1], sys.argv[2])
    print(json.dumps(result, indent=2, ensure_ascii=False))

File: .claude/skills/crucible-spec-workspace/grade_spec.ts
#!/usr/bin/env bun

import { existsSync } from "node:fs";

type Assertion = {
  name: string;
  text: string;
  type?: string;
  pattern?: string;
  heading?: string;
};

type Metadata = {
  eval_id: string;
  eval_name: string;
  assertions: Assertion[];
};

function fail(message: string): never {
  throw new Error(message);
}

async function gradeSpec(specPath: string, metadataPath: string): Promise<Record<string, unknown>> {
  const spec = existsSync(specPath) ? await Bun.file(specPath).text() : "";
  const metadata = JSON.parse(await Bun.file(metadataPath).text()) as Metadata;

  const results = metadata.assertions.map((assertion) => {
    let passed = false;
    let evidence = "";

    if (assertion.type === "regex" && assertion.pattern) {
      const matches = spec.match(new RegExp(assertion.pattern, "g")) ?? [];
      passed = matches.length > 0;
      evidence = passed
        ? `Found ${matches.length} matches: ${matches.slice(0, 3).join(", ")}`
        : `No match for pattern: ${assertion.pattern}`;
    } else if (assertion.type === "heading_exists" && assertion.heading) {
      const pattern = new RegExp(`^#{1,4}\\s+.*${assertion.heading.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}.*$`, "gim");
      const matches = spec.match(pattern) ?? [];
      passed = matches.length > 0;
      evidence = passed ? `Found heading: ${matches[0]}` : `No heading containing '${assertion.heading}'`;
    } else if (assertion.type === "content_check") {
      if (assertion.text.includes("알림 타입") || assertion.text.toLowerCase().includes("notification")) {
        const keywords = ["주문", "메시지", "공지"];
        const found = keywords.filter((keyword) => spec.includes(keyword));
        passed = found.length >= 3;
        evidence = `Found ${found.length}/3 notification types: ${found.join(", ")}`;
      }
    } else if (assertion.type === "composite") {
      const required = ["Problem", "Journey", "Feature", "Constraint", "Acceptance", "Data Model"];
      const found = required.filter((section) => new RegExp(`^#{1,4}\\s+.*${section}.*$`, "gim").test(spec));
      const missing = required.filter((section) => !found.includes(section));
      passed = missing.length === 0;
      evidence =
        missing.length === 0
          ? `All ${required.length} required sections present`
          : `Found ${found.length}/${required.length} sections. Missing: ${missing.join(", ")}`;
    }

    return {
      text: assertion.text,
      passed,
      evidence,
    };
  });

  const passCount = results.filter((result) => result.passed).length;
  return {
    eval_id: metadata.eval_id,
    eval_name: metadata.eval_name,
    expectations: results,
    pass_rate: results.length > 0 ? passCount / results.length : 0,
  };
}

async function main(): Promise<number> {
  const args = process.argv.slice(2);
  if (args.length !== 2) {
    console.error("Usage: grade_spec.ts <spec_file> <metadata_file>");
    return 1;
  }

  const [specPath, metadataPath] = args;
  if (!specPath || !metadataPath) {
    fail("spec_path와 metadata_path가 필요합니다.");
  }

  console.log(JSON.stringify(await gradeSpec(specPath, metadataPath), null, 2));
  return 0;
}

await main().then((code) => process.exit(code)).catch((error: unknown) => {
  console.error(error instanceof Error ? error.message : String(error));
  process.exit(1);
});
